server/src/com/thoughtworks/go/server/websocket/browser/subscription/JobStatusChangeSubscriptionHandler.java

```diff
@@ -0,0 +1,121 @@
+    }
+
+    @Override
+    public void start(SubscriptionMessage message, BrowserWebSocket socket) throws Exception {
+        message.start(socket, this);
+    }
+
+    @Override
+    public boolean isAuthorized(SubscriptionMessage message, SecurityService securityService, Username currentUser) {
+        return message.isAuthorized(securityService, currentUser);
+    }
+
+    @Override
+    public Class getType() {
+        return JobStatusChange.class;
+    }
+
+    public void sendCurrentJobInstance(JobIdentifier jobIdentifier, BrowserWebSocket socket) throws IOException {
+        sendJobInstance(getjobStatusJson(jobIdentifier), socket);
+    }
+
+    public JobInstance getjobInstance(JobIdentifier identifier) {
+        JobConfigIdentifier jobConfigIdentifier = new JobConfigIdentifier(identifier.getPipelineName(), identifier.getStageName(), identifier.getBuildName());
+        return cache.currentJob(jobConfigIdentifier);
+    }
+
+
+    public List getjobStatusJson(JobIdentifier identifier) {
+        JobConfigIdentifier jobConfigIdentifier = new JobConfigIdentifier(identifier.getPipelineName(), identifier.getStageName(), identifier.getBuildName());
+        JobInstance jobInstance = cache.currentJob(jobConfigIdentifier);
+        JobStatusJsonPresentationModel presenter = new JobStatusJsonPresentationModel(jobInstance,
+                agentService.findAgentObjectByUuid(jobInstance.getAgentUuid()),
+                stageService.getBuildDuration(identifier.getPipelineName(), identifier.getStageName(), jobInstance));
+        return createBuildInfo(presenter);
+    }
+
+    private List createBuildInfo(JobStatusJsonPresentationModel presenter) {
+        Map<String, Object> info = new LinkedHashMap<>();
+        info.put("building_info", presenter.toJsonHash());
+        List jsonList = new ArrayList();
+        jsonList.add(info);
+        return jsonList;
+    }
+
+    public void registerJobStateChangeListener(JobIdentifier jobIdentifier, BrowserWebSocket socket) {
+        jobStates.put(jobIdentifier, JobState.Unknown);
+        jobInstanceService.registerJobStateChangeListener(new JobStatusListener() {
+            @Override
+            public void jobStatusChanged(JobInstance job)  {
+                if(!jobStates.get(jobIdentifier).getStatus().equals(job.getState())) {
+                    try {
+                        sendJobInstance(getjobStatusJson(jobIdentifier), socket);
+                        jobStates.put(jobIdentifier, job.getState());
+                        if(job.isCompleted()) {
+                            socket.close();
+                        }
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            }
+        });
+    }
+
+    private void sendJobInstance(List json, BrowserWebSocket socket) throws IOException {
+        socket.send(ByteBuffer.wrap(new Gson().toJson(json).getBytes()));
```

Comment:
Wrap the response with type of event, so only the specific event listener will parse the response